File: samplejava/src/main/java/com/example/chattutorialjava/ChannelActivity3.java
package com.example.chattutorialjava;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.activity.OnBackPressedCallback;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.getstream.sdk.chat.view.MessageListView;
import com.getstream.sdk.chat.view.messageinput.MessageInputView;
import com.getstream.sdk.chat.viewmodel.MessageInputViewModel;
import com.getstream.sdk.chat.viewmodel.MessageInputViewModelBinding;
import com.getstream.sdk.chat.viewmodel.messages.MessageListViewModel;
import com.getstream.sdk.chat.viewmodel.messages.MessageListViewModelBinding;

import java.util.LinkedList;
import java.util.List;

import io.getstream.chat.android.client.models.Channel;
import io.getstream.chat.android.client.models.User;
import io.getstream.chat.android.livedata.ChatDomain;
import io.getstream.chat.android.livedata.controller.ChannelController;
import kotlin.Unit;
import kotlin.jvm.functions.Function0;

public class ChannelActivity3 extends AppCompatActivity {

    private final static String CID_KEY = "key:cid";
    private final static int MESSAGE_LIMIT = 30;

    public static Intent newIntent(Context context, Channel channel) {
        final Intent intent = new Intent(context, ChannelActivity3.class);
        intent.putExtra(CID_KEY, channel.getCid());
        return intent;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_channel_3);
        final MessageListView messageListView = findViewById(R.id.messageListView);
        final ProgressBar progressBar = findViewById(R.id.progressBar);
        final TextView channelHeaderView = findViewById(R.id.channelHeaderView);
        final MessageInputView messageInputView = findViewById(R.id.messageInputView);

        messageListView.setAttachmentViewHolderFactory(new MyAttachmentViewHolderFactory());

        final String cid = getIntent().getStringExtra(CID_KEY);
        final ViewModelProvider viewModelProvider = new ViewModelProvider(this, new ChannelViewModelsFactory(cid));

        final MessageListViewModel messageListViewModel = viewModelProvider.get(MessageListViewModel.class);
        MessageListViewModelBinding.bind(messageListViewModel, messageListView, this);
        messageListViewModel.getState().observe(this, state -> {
            if (state instanceof MessageListViewModel.State.Loading) {
                progressBar.setVisibility(View.VISIBLE);
            } else if (state instanceof MessageListViewModel.State.Result) {
                progressBar.setVisibility(View.GONE);
            } else if (state instanceof MessageListViewModel.State.NavigateUp) {
                finish();
            }
        });

<<<<<<< HEAD
        String nobodyTypingText = "nobody is typing";
        channelHeaderView.setText(nobodyTypingText);
        ChatDomain.instance().getUseCases().getWatchChannel().invoke(cid, MESSAGE_LIMIT).enqueue(result -> {
            if (result.isSuccess()) {
                ChannelController channelController = result.data();
                runOnUiThread(() -> channelController.getTyping().observe(this, users -> {
                    String typing = nobodyTypingText;
                    if (!users.isEmpty()) {
                        List<String> names = new LinkedList<>();
                        for (User user : users) {
                            names.add((String) user.getExtraData().get("name"));
                        }
                        typing = "typing: " + TextUtils.join(", ", names);
                    }
                    channelHeaderView.setText(typing);
                }));
            }
            return Unit.INSTANCE;
=======
        ChannelController channelController = ChatClient.instance().channel(cid);
        MutableLiveData<Set<String>> currentlyTyping = new MutableLiveData<>(new HashSet<>());
        channelController.subscribeFor(
                new Class[]{TypingStartEvent.class, TypingStartEvent.class},
                event -> {
                    if (event instanceof TypingStartEvent) {
                        User user = ((TypingStartEvent) event).getUser();
                        String name = (String) user.getExtraData().get("name");
                        Set<String> typingCopy = currentlyTyping.getValue();
                        typingCopy.add(name);
                        currentlyTyping.postValue(typingCopy);
                    } else if (event instanceof TypingStopEvent) {
                        User user = ((TypingStopEvent) event).getUser();
                        String name = (String) user.getExtraData().get("name");
                        Set<String> typingCopy = currentlyTyping.getValue();
                        typingCopy.remove(name);
                        currentlyTyping.postValue(typingCopy);
                    }
                    return null;
                });
        currentlyTyping.observe(this, users -> {
            String typing = "nobody is typing";
            if (!users.isEmpty()) {
                typing = "typing: " + TextUtils.join(", ", users);
            }
            channelHeaderView.setText(typing);
>>>>>>> 23067e524310df9e383d921476e1f7625c2b04df
        });

        final MessageInputViewModel messageInputViewModel = viewModelProvider.get(MessageInputViewModel.class);
        MessageInputViewModelBinding.bind(messageInputViewModel, messageInputView, this);
        messageListViewModel.getMode().observe(this, mode -> {
            if (mode instanceof MessageListViewModel.Mode.Thread) {
                messageInputViewModel.setActiveThread(((MessageListViewModel.Mode.Thread) mode).getParentMessage());
            } else if (mode instanceof MessageListViewModel.Mode.Normal) {
                messageInputViewModel.resetThread();
            }
            messageListView.setOnMessageEditHandler((message) -> {
                messageInputViewModel.getEditMessage().postValue(message);
                return Unit.INSTANCE;
            });
        });

        Function0<Unit> onBackClick = () -> {
            messageListViewModel.onEvent(MessageListViewModel.Event.BackButtonPressed.INSTANCE);
            return Unit.INSTANCE;
        };

        getOnBackPressedDispatcher().addCallback(this, new OnBackPressedCallback(true) {
            @Override
            public void handleOnBackPressed() {
                onBackClick.invoke();
            }
        });
    }
}

File: app/src/main/java/com/example/chattutorial/ChannelActivity3.kt
package com.example.chattutorial

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.getstream.sdk.chat.view.common.visible
import com.getstream.sdk.chat.viewmodel.MessageInputViewModel
import com.getstream.sdk.chat.viewmodel.bindView
import com.getstream.sdk.chat.viewmodel.messages.MessageListViewModel
import com.getstream.sdk.chat.viewmodel.messages.bindView
import io.getstream.chat.android.client.ChatClient
import io.getstream.chat.android.client.events.TypingStartEvent
import io.getstream.chat.android.client.events.TypingStopEvent
import io.getstream.chat.android.client.models.Channel
import kotlinx.android.synthetic.main.activity_channel_3.*

class ChannelActivity3 : AppCompatActivity(R.layout.activity_channel_3) {

    private val cid: String by lazy {
        intent.getStringExtra(CID_KEY)!!
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val viewModelProvider = ViewModelProvider(this, ChannelViewModelsFactory(cid))

        messageListView.setViewHolderFactory(MyMessageViewHolderFactory())
        val messagesViewModel = viewModelProvider.get(MessageListViewModel::class.java)
            .apply {
                bindView(messageListView, this@ChannelActivity3)
                state.observe(
                    this@ChannelActivity3,
                    {
                        when (it) {
                            is MessageListViewModel.State.Loading -> progressBar.visible(true)
                            is MessageListViewModel.State.Result -> progressBar.visible(false)
                            is MessageListViewModel.State.NavigateUp -> finish()
                        }
                    }
                )
            }

        val channelController = ChatClient.instance().channel(cid)
        val currentlyTyping = MutableLiveData<Set<String>>(HashSet())

        channelController.events().subscribe {
            val typing = currentlyTyping.value ?: setOf()
            val typingCopy: MutableSet<String> = typing.toMutableSet()
            when (it) {
                is TypingStartEvent -> {
                    val name = it.user.extraData["name"]!! as String
                    typingCopy += name
                    currentlyTyping.postValue(typingCopy)
                }
                is TypingStopEvent -> {
                    val name = it.user.extraData["name"]!! as String
                    typingCopy -= name
                    currentlyTyping.postValue(typingCopy)
                }
            }
        }

        val typingObserver = Observer<Set<String>> { users ->
            var typing = "nobody is typing"
            if (users.isNotEmpty()) {
                typing = "typing: " + users.joinToString(", ")
            }
            channelHeaderView.text = typing
        }
        currentlyTyping.observe(this, typingObserver)

        viewModelProvider.get(MessageInputViewModel::class.java).apply {
            bindView(messageInputView, this@ChannelActivity3)
            messagesViewModel.mode.observe(
                this@ChannelActivity3,
                {
                    when (it) {
                        is MessageListViewModel.Mode.Thread -> setActiveThread(it.parentMessage)
                        is MessageListViewModel.Mode.Normal -> resetThread()
                    }
                }
            )
            messageListView.setOnMessageEditHandler {
                editMessage.postValue(it)
            }
        }
        val backButtonHandler = {
            messagesViewModel.onEvent(MessageListViewModel.Event.BackButtonPressed)
        }

        onBackPressedDispatcher.addCallback(
            this,
            object : OnBackPressedCallback(true) {
                override fun handleOnBackPressed() {
                    backButtonHandler()
                }
            }
        )
    }

    companion object {
        private const val CID_KEY = "key:cid"

        fun newIntent(context: Context, channel: Channel) =
            Intent(context, ChannelActivity3::class.java).apply {
                putExtra(CID_KEY, channel.cid)
            }
    }
}
